### PR TITLE
fix: fix performance regression

### PR DESCRIPTION
**Pull request by Shadow53**
Rough stats for these changes. Timing is based on parsing the sample calibrations file.

| Before regression  | Before Fix                  | After fix   | After, with `lto=fat,codegen-units=1`|
|  ----------------- | --------------------- | --------- | ------------------------------------- |
| ~40ms                    | ~30,000-60,000ms   |  ~230ms | ~185 ms                                                 |

Parsing is generally more expensive now because we are tracking location information while parsing, which adds some overhead (though not as much after these changes).

All of the `pub fn parse_*` functions were already crate-private by being marked `pub` in a private module. I just made them `pub(crate)` for clarity, since I was modifying return types already.

Edit: thousands of milliseconds, not thousands of seconds

### CODE REVIEW

File: src/parser/macros.rs
Comment (notmgsk) [inline src/parser/macros.rs:15]: here or a future issue?

Comment (notmgsk) [inline src/parser/macros.rs:15]:
```suggestion
```

**Comment by notmgsk** [comment]
Before merging we should:
- check it doesn't have any incompatibilities with something that uses this crate

**Comment by genos** [comment]
🔥

File: src/parser/error/internal.rs
Comment (kalzoo) [inline src/parser/error/internal.rs:66]: TODO without linked issue => PR should be draft

File: src/parser/error/internal.rs
Comment (kalzoo) [inline src/parser/error/internal.rs:20]: Maybe add a little bit about _why_?

- why there are two error types
- why the conversion should happen when it does

File: Cargo.toml
Comment (kalzoo) [inline Cargo.toml:34]: This is still benching the `parser`, it's just an additional fixture, can we include it in the same command?

**Review by kalzoo — COMMENTED** [review]
There and back again, @Shadow53 ! good find and fix. Approved pending fixes for my comments.

**Comment by kalzoo** [comment]
(also, please create an additional issue to drive the time back down with any ideas you have of where to start looking)

**Comment by Shadow53** [comment]
> (also, please create an additional issue to drive the time back down with any ideas you have of where to start looking)

Created #115 to address this.

**Review by notmgsk — APPROVED** [review]
My test program still takes ~5s, but that is much less frustrating than ~30s.

**Comment by Shadow53** [comment]
> My test program still takes ~5s, but that is much less frustrating than ~30s.

Leaving a note here for documentation:

The numbers in the OP are using default benchmarking settings, which are equivalent to using release mode. If you tell it to use the `dev` profile instead...

```
$ cargo bench --profile=dev
# ... snip ...
     Running benches/parser.rs (target/debug/deps/parser-d2d0ad29d261c397)
Gnuplot not found, using plotters backend
Benchmarking calibration file/Sample calibration file: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 1173.2s, or reduce sample count to 10.
Benchmarking calibration file/Sample calibration file: Collecting 100 samples in estimated 1173.2 s (100 iterations)
```

File: src/parser/error/internal.rs
Comment (kalzoo) [inline src/parser/error/internal.rs:22]:
```suggestion
/// `InternalError` is an intermediate error that contains a reference to the failed input
```

**Review by kalzoo — APPROVED** [review]
Looks good! Small typo fix, otherwise 🚢

**Comment by kalzoo** [comment]
> > My test program still takes ~5s, but that is much less frustrating than ~30s.
> 
> Leaving a note here for documentation:
> 
> The numbers in the OP are using default benchmarking settings, which are equivalent to using release mode. If you tell it to use the `dev` profile instead...
> 
> ```
> $ cargo bench --profile=dev
> # ... snip ...
>      Running benches/parser.rs (target/debug/deps/parser-d2d0ad29d261c397)
> Gnuplot not found, using plotters backend
> Benchmarking calibration file/Sample calibration file: Warming up for 3.0000 s
> Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 1173.2s, or reduce sample count to 10.
> Benchmarking calibration file/Sample calibration file: Collecting 100 samples in estimated 1173.2 s (100 iterations)
> ```
> 
> 1173.2s / 100 iterations = ~12s to parse the sample calibration file.

Good to note - I had not realized that. This bumps up the priority of the extended fix here (#115).

**Pull request merged**
